### PR TITLE
[Merged by Bors] - feat: add Fin.cons_mem_piFinset_iff

### PR DESCRIPTION
**Pull request by BoltonBailey**
Adds a lemma about when `Fin.cons` is an element of `piFinset`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fintype/Pi.lean [hunk 1]
```diff
@@ -109,3 +110,18 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+@[simp]
+lemma Fin.cons_mem_piFinset_iff {α} {n : ℕ} (p : (Fin n → α)) (x : α) (S : Finset α) :
+    Fin.cons x p ∈ Fintype.piFinset (fun _ => S) ↔ x ∈ S ∧ p ∈ Fintype.piFinset (fun _ => S) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:116]: Can you generalize `fun _ => S` here to `S : Fin n → Finset α`? I think that gives
```suggestion
    Fin.cons x p ∈ Fintype.piFinset (fun _ => S) ↔ x ∈ S 0 ∧ p ∈ Fintype.piFinset (fun i => S i.succ) := by
```

File: Mathlib/Data/Fintype/Pi.lean [hunk 2]
```diff
@@ -109,3 +110,18 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+@[simp]
+lemma Fin.cons_mem_piFinset_iff {α} {n : ℕ} (p : (Fin n → α)) (x : α) (S : Finset α) :
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:115]: Can you add the `Fin.snoc` lemma too for good measure?

File: Mathlib/Data/Fintype/Pi.lean (hunk 1 above)
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean:116]: Ok will do

Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: Actually I thought it looked asymmetrical to have cons on the LHS but not the right, so I split into two versions, one with cons and one without.

File: Mathlib/Data/Fintype/Pi.lean [hunk 3]
```diff
@@ -109,3 +110,26 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+lemma Fin.succ_mem_piFinset_iff {α} {n : ℕ} (p : (Fin (n + 1) → α)) (S : Fin (n+1) → Finset α) :
+    p ∈ Fintype.piFinset (fun i => S i)
+      ↔
+    p 0 ∈ S 0 ∧ (Fin.tail p) ∈ Fintype.piFinset (Fin.tail S) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:117]:
```suggestion
    p ∈ Fintype.piFinset S ↔ p 0 ∈ S 0 ∧ Fin.tail p ∈ Fintype.piFinset (Fin.tail S) := by
```

File: Mathlib/Data/Fintype/Pi.lean (hunk 2 above)
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean]: Ok done, but I had to add API to do it, so the PR now touches Fin.Basic

File: Mathlib/Data/Fintype/Pi.lean [hunk 4]
```diff
@@ -109,3 +110,26 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+lemma Fin.succ_mem_piFinset_iff {α} {n : ℕ} (p : (Fin (n + 1) → α)) (S : Fin (n+1) → Finset α) :
```
[the remark this answers is not in the document]
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean:114]: I'm getting a lot of "failed to synthesize instance Membership"

[the remark this answers is not in the document]
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean:114]: Err, wait I mistyped, sorry

[the remark this answers is not in the document]
Comment (BoltonBailey) [reply Mathlib/Data/Fintype/Pi.lean:114]: Yeah it goes through

**Comment by alexjbest** [comment]
looks like you committed things that are in lake-packages as submodules somehow please remove them

File: Mathlib/Data/Fintype/Pi.lean [hunk 5]
```diff
@@ -109,3 +110,49 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+lemma Fin.succ_mem_piFinset_iff {n : ℕ} {α : Fin (n + 1) → Type*} (p : ∀ i, α i)
+  (S : ∀ i, Finset (α i)) :
+    p ∈ Fintype.piFinset S ↔ p 0 ∈ S 0 ∧ (Fin.tail p) ∈ Fintype.piFinset (Fin.tail S) := by
```
Comment (alexjbest) [inline Mathlib/Data/Fintype/Pi.lean:116]:
```suggestion
    p ∈ Fintype.piFinset S ↔ p 0 ∈ S 0 ∧ Fin.tail p ∈ Fintype.piFinset (Fin.tail S) := by
```
I dont think these brackets are needed

File: Mathlib/Data/Fin/Basic.lean [hunk 6]
```diff
@@ -1877,6 +1877,18 @@ theorem coe_ofNat_eq_mod (m n : ℕ) [NeZero m] :
   rfl
 #align fin.coe_of_nat_eq_mod Fin.coe_ofNat_eq_mod
 
+theorem eq_castSucc_or_eq_last {n : Nat} : ∀ i : Fin (n + 1),
+    (∃ j : Fin n, i = j.castSucc) ∨ i = (last n) := by
```
Comment (alexjbest) [inline Mathlib/Data/Fin/Basic.lean:1881]:
```suggestion
    (∃ j : Fin n, i = j.castSucc) ∨ i = last n := by
```
shouldnt be needed

File: Mathlib/Data/Fintype/Pi.lean [hunk 7]
```diff
@@ -109,3 +110,49 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
...
+    constructor
+    · exact fun a_1 ↦ h1 (Fin.castSucc a_1)
+    · exact h1 (Fin.last n)
+  · intro ⟨h1, h2⟩ a1
+    rcases Fin.eq_castSucc_or_eq_last a1 with ⟨j, rfl⟩ | rfl
+    · apply h1
+    · exact h2
+
+@[simp]
+lemma Fin.snoc_mem_piFinset_cons_iff {α} {n : ℕ} (p : Fin n → α) (x : α)
```
Comment (jcommelin) [inline Mathlib/Data/Fintype/Pi.lean:153]: Did you mean
```suggestion
lemma Fin.snoc_mem_piFinset_snoc_iff {α} {n : ℕ} (p : Fin n → α) (x : α)
```

File: Mathlib/Data/Fintype/Pi.lean [hunk 8]
```diff
@@ -109,3 +110,49 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
     (Finset.univ.pi fun a : α => (Finset.univ : Finset (β a))) = Finset.univ := by
   ext; simp
 #align finset.univ_pi_univ Finset.univ_pi_univ
+
+lemma Fin.succ_mem_piFinset_iff {n : ℕ} {α : Fin (n + 1) → Type*} (p : ∀ i, α i)
+  (S : ∀ i, Finset (α i)) :
```
Comment (ocfnash) [inline Mathlib/Data/Fintype/Pi.lean:115]: I think `succ` should appear later in the name, and we should use the new notation for dependent functions (which I find much easier to read). Also we should consistent intent by 4 spaces.
```suggestion
lemma Fin.mem_piFinset_succ_iff {n : ℕ} {α : Fin (n + 1) → Type*} (p : (i : Fin (n + 1)) → α i)
    (S : (i : Fin (n + 1)) → Finset (α i)) :
```
(and similarly in various lemmas below)

File: Mathlib/Data/Fin/Basic.lean [hunk 9]
```diff
@@ -1876,6 +1876,18 @@ theorem coe_ofNat_eq_mod (m n : ℕ) [NeZero m] :
   rfl
 #align fin.coe_of_nat_eq_mod Fin.coe_ofNat_eq_mod
 
+theorem eq_castSucc_or_eq_last {n : Nat} : ∀ i : Fin (n + 1),
+    (∃ j : Fin n, i = j.castSucc) ∨ i = last n := by
+  intro ⟨i, hi⟩
```
Comment (ocfnash) [inline Mathlib/Data/Fin/Basic.lean:1881]:
```suggestion
theorem eq_castSucc_or_eq_last {n : Nat} (i : Fin (n + 1)) :
    (∃ j : Fin n, i = j.castSucc) ∨ i = last n := by
  obtain ⟨i : ℕ, hi : i < n + 1⟩ := i
```

**Comment by BoltonBailey** [comment]
For the show_term golfing, I am confused if you want me to do that for the blocks with rcases. Doing that seems to make the proof longer.

File: Mathlib/Data/Fin/Basic.lean [hunk 10]
```diff
@@ -1780,6 +1780,18 @@ theorem coe_ofNat_eq_mod (m n : ℕ) [NeZero m] :
...
+  obtain ⟨i : ℕ, hi : i < n + 1⟩ := i
+  by_cases h : i < n
+  · left
+    use ⟨i, by assumption⟩
+    simp only [castSucc_mk]
+  · right
+    apply eq_of_val_eq
+    simp only [val_last]
+    exact Nat.eq_of_lt_succ_of_not_lt hi h
+
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:1794]: This looks like an induction principle in disguise to me.

Comment (eric-wieser) [reply Mathlib/Data/Fin/Basic.lean:1794]: Indeed, this is `Fin.reverseInduction`

Comment (eric-wieser) [reply Mathlib/Data/Fin/Basic.lean:1794]: You can prove this with
```
  induction i using reverseInduction with
  | last => right; rfl
  | cast n => left; exact ⟨_, rfl⟩
```
but instead you should just delete this lemma.

File: Mathlib/Data/Fintype/Pi.lean [hunk 11]
```diff
@@ -109,3 +110,43 @@ theorem Finset.univ_pi_univ {α : Type*} {β : α → Type*} [DecidableEq α] [F
...
+    · assumption
+    · apply ha12
+
+@[simp]
+lemma Fin.cons_mem_piFinset_cons_iff {α} {n : ℕ} (p : Fin n → α) (x : α)
+    (S₀ : Finset α) (Sᵢ : Fin n → Finset α) :
+    Fin.cons x p ∈ Fintype.piFinset (Fin.cons (α := fun _ => Finset α) S₀ Sᵢ)
+      ↔
+    x ∈ S₀ ∧ p ∈ Fintype.piFinset Sᵢ := by
+  simp_rw [Fin.mem_piFinset_succ_iff, cons_zero, tail_cons]
```
Comment (eric-wieser) [inline Mathlib/Data/Fintype/Pi.lean:131]: This generalizes:
```suggestion
@[simp]
lemma Fin.cons_mem_piFinset_cons_iff {n : ℕ} {α : Fin (n + 1) → Type*}
    (x : α 0) (xs : (i : Fin _) → α i.succ)
    (S₀ : Finset (α 0)) (Sᵢ : (i : Fin n) → Finset (α i.succ)) :
    Fin.cons x xs ∈ Fintype.piFinset (Fin.cons S₀ Sᵢ) ↔ x ∈ S₀ ∧ xs ∈ Fintype.piFinset Sᵢ := by
  simp_rw [Fin.mem_piFinset_succ_iff, cons_zero, tail_cons]
```

**Review by eric-wieser — APPROVED** [review]
bors d+

I pushed changes that address all my suggestions; if they look ok to you, please merge; otherwise revert the ones you don't like and ping me.

**Pull request closed without merging**